File: app/src/main/java/com/chandilsachin/diettracker/MainActivity.java
package com.chandilsachin.diettracker;

import android.content.Intent;
import android.os.AsyncTask;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.chandilsachin.diettracker.mvp.view.ProgressDialog;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class MainActivity extends AppCompatActivity {

    private DatabaseManager dbManager;
    private MainPresenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        dbManager = DatabaseManager.getInstance(this);
        presenter = new MainPresenter();
        initViews();
        setEvents();
        setUpDatabase();
        prepareDietList();
    }

    private void setUpDatabase(){
        if(!new InitPreferences(this).hasDataLoaded()){
            new SetUpFoodDatabase().execute();
        }
    }

    private void prepareDietList(){
        new Thread(new Runnable() {
            @Override
            public void run() {
                final ArrayList<Food> list = presenter.getDietList(MainActivity.this);
                MainActivity.this.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        bindDataToList(list);
                    }
                });
            }
        }).start();

    }


    private void bindDataToList(ArrayList<Food> list){
        DietListAdapter adapter = new DietListAdapter(this, list);
        recyclerViewDietList.setAdapter(adapter);
    }

    private RelativeLayout relativeLayoutAddBreakfast;
    private CardView linearLayoutBreakfast;
    private RecyclerView recyclerViewDietList;
    public void initViews(){
        linearLayoutBreakfast = (CardView) findViewById(R.id.linearLayoutBreakfast);
        relativeLayoutAddBreakfast = (RelativeLayout) linearLayoutBreakfast.findViewById(R.id.relativeLayoutAddFood);
        recyclerViewDietList = (RecyclerView) findViewById(R.id.recyclerViewDietList);
        recyclerViewDietList.setLayoutManager(new LinearLayoutManager(this));
    }

    public void setEvents(){
        relativeLayoutAddBreakfast.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, AddFoodActivity.class);
                startActivityForResult(intent, AddFoodActivity.CODE_FOOD_SELECTION);
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode){
            case AddFoodActivity.CODE_FOOD_SELECTION:
                int selectedFood = data.getIntExtra(AddFoodActivity.SELECTED_INDEX, -1);
                dbManager.saveFood(selectedFood, 1f, Calendar.getInstance().getTime());
                Toast.makeText(this, "Selected Item: "+ selectedFood, Toast.LENGTH_SHORT).show();
                prepareDietList();
                break;
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

    }

    class SetUpFoodDatabase extends AsyncTask<Void, Integer, Void>{

        android.app.ProgressDialog dialog;
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            dialog = ProgressDialog.show(MainActivity.this, "Preparing Database", "Preparing food database...", true);
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            dialog.dismiss();
            new InitPreferences(MainActivity.this).setDataHasLoaded(true);
        }

        @Override
        protected Void doInBackground(Void... params) {
            try {
                ArrayList<Food> list = new AddFoodPresenter().readFoodItemsFile(MainActivity.this);
                dbManager.addFoodList(list);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }
    }
}

File: app/src/main/java/com/chandilsachin/diettracker/MainActivity.kt
package com.chandilsachin.diettracker

import android.arch.lifecycle.LifecycleActivity
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.CardView
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.widget.RelativeLayout
import com.chandilsachin.diettracker.Food
import com.chandilsachin.diettracker.mvp.view.ProgressDialog
import kotlinx.android.synthetic.main.layout_diary_page.*
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import java.io.IOException
import java.util.*

class MainActivity : LifecycleActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initViews()
        setEvents()
        setUpDatabase()
        prepareDietList()
    }

    private fun setUpDatabase() {
        if (!InitPreferences(this).hasDataLoaded()) {
            setUpFoodDatabase()
        }
    }

    private fun prepareDietList() {
        val model = ViewModelProviders.of(this).get(MainActivityModel::class.java)
        model.userLiveData.observe(this, android.arch.lifecycle.Observer<ArrayList<Food>> { list->
            list?.let {
                if(list.size > 0){
                    bindDataToList(list)
                }
            }

        })
        /*doAsync {
            //val list = FoodDatabase.getInstance(baseContext)?.getFood(Calendar.getInstance().time.time)
            val list = DatabaseManager.getInstance(baseContext).getFoodOnDate(Calendar.getInstance().time);
            uiThread {
                if (list != null)
                    bindDataToList(list)
            }
        }*/
    }


    private fun bindDataToList(list: ArrayList<Food>) {
        val adapter = DietListAdapter(baseContext, list)
        recyclerViewDietList.adapter = adapter
    }

    private var relativeLayoutAddBreakfast: RelativeLayout? = null
    private var linearLayoutBreakfast: CardView? = null
    fun initViews() {
        linearLayoutBreakfast = findViewById(R.id.linearLayoutBreakfast) as CardView
        relativeLayoutAddBreakfast = linearLayoutBreakfast!!.findViewById(R.id.relativeLayoutAddFood) as RelativeLayout
        recyclerViewDietList.layoutManager = LinearLayoutManager(this)
    }

    fun setEvents() {
        relativeLayoutAddBreakfast!!.setOnClickListener {
            val intent = Intent(this@MainActivity, AddFoodActivity::class.java)
            startActivityForResult(intent, AddFoodActivity.CODE_FOOD_SELECTION)
        }
    }

    fun setUpFoodDatabase(){

        var dialog = ProgressDialog.show(this@MainActivity, "Preparing Database", "Preparing food database...", true)
        doAsync {
            try {
                val list = AddFoodPresenter().readFoodItemsFile(this@MainActivity)
                //FoodDatabase.getInstance(baseContext)?.addFoodList(list)
                DatabaseManager.getInstance(baseContext).addFoodList(list);
            } catch (e: IOException) {
                e.printStackTrace()
            }
            uiThread {
                InitPreferences(this@MainActivity).setDataHasLoaded(true)
                dialog?.dismiss()
            }
        }
    }
}